simpegEM1D/EM1D.py

```diff
@@ -262,104 +246,136 @@ def forward(self, m, output_type='response'):
-                    dhz = self.hz_kernel_circular_loop(
-                        self.YBASE/a, f[ifreq], n_layer,
+                    hz[ifreq, :, :] = self.hz_kernel_circular_loop(
+                        lambd[ifreq, :], f[ifreq], n_layer,
                         sig, chi, depth, h, z, I, a,
                         flag, output_type=output_type
                     )
-                    dHzFHT_dsig[ifreq, :] = np.dot(dhz, self.WT1)/a
+                # ??? D.W. lambd here was self.YBASE/a?
+                # ??? D.W. The result after DLF was divided by a?
+
+                PJ = (hz0, hz, hz0)  # PJ1
+
             else:
                 raise Exception("Src options are only VMD or CircularLoop!!")
 
-            return dHzFHT_dsig
-
         elif output_type == 'sensitivity_height':
-            dHzFHT_dh = np.empty((n_frequency, 1), dtype=complex)
-            dhz = np.empty(nfilt, complex)
+
+            # for simulation
+            hz = np.empty((n_frequency, nfilt), complex)
+            hz0 = np.zeros((n_frequency, nfilt), complex)  # For zero kernels
+
             if self.survey.src_type == 'VMD':
-                r = self.survey.offset
                 for ifreq in range(n_frequency):
                     sig = self.sigma_cole(f[ifreq])
-                    dhz = self.hz_kernel_vertical_magnetic_dipole(
-                        self.YBASE/r[ifreq], f[ifreq], n_layer,
+                    hz[ifreq, :] = self.hz_kernel_vertical_magnetic_dipole(
+                        lamb[ifreq, :], f[ifreq], n_layer,
                         sig, chi, depth, h, z,
                         flag, output_type=output_type
                     )
-                    dHzFHT_dh[ifreq] = np.dot(dhz, self.WT0)/r[ifreq]
+
+                PJ = (hz, hz0, hz0)  # PJ0
 
             elif self.survey.src_type == 'CircularLoop':
                 I = self.survey.I
                 a = self.survey.a
                 for ifreq in range(n_frequency):
                     sig = self.sigma_cole(f[ifreq])
-                    dhz = self.hz_kernel_circular_loop(
-                        self.YBASE/a, f[ifreq], n_layer,
+                    hz = self.hz_kernel_circular_loop(
+                        lambd[ifreq, :], f[ifreq], n_layer,
                         sig, chi, depth, h, z, I, a,
                         flag, output_type=output_type
                     )
-                    dHzFHT_dh[ifreq] = np.dot(dhz, self.WT1)/a
+                # ??? D.W. lambd here was self.YBASE/a?
+                # ??? D.W. The result after DLF was divided by a?
+
+                PJ = (hz0, hz, hz0)  # PJ1
+
             else:
                 raise Exception("Src options are only VMD or CircularLoop!!")
 
-            return dHzFHT_dh
+        # Carry out Hankel DLF
+        # ab=66 => 33 (vertical magnetic src and rec)
+        HzFHT = dlf(PJ, lambd, r, self.fhtfilt, self.hankel_pts_per_dec,
+                    factAng=np.full_like(r, 1.), ab=33)
```

Comment:
so, for `CircularLoop` I need to put `a` which is the radius of the loop rather than `r`
In addition, `VMD` and `CircularLoop` use different kernel: `j1` and `j0`, how can I make distinctions here? (e.g. `ab` value for `CircularLoop`?)

Reply:
A circular loop is in the end a vertical magnetic field, so you can just use `ab=33` for both. It should even work leaving `ab` out in your case.

Reply:
"I need to put `a` in...` I don't think so. The area affects the strength of your signal. But the Hankel transform depends on offset, not on area.